config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java

```diff
@@ -112,15 +112,24 @@
                 String configPropName = originalNames.get(beanProp.getName());
                 // Is the property key missing in the config?
                 if (configPropName == null) {
-                    // If so, continue if the field is marked as @{link Optional}
-                    if (isOptionalProperty(clazz, beanProp)) {
-                        continue;
+                    // If the field is not marked as @{link Optional}...
+                    if (!isOptionalProperty(clazz, beanProp)) {
+                        // raise a {@link Missing} exception right here
+                        throw new ConfigException.Missing(beanProp.getName());
+                    }
+                } else {
+                    try {
+                        Object unwrapped = getValue(
+                          clazz, parameterType, parameterClass, config, configPropName);
+                        setter.invoke(bean, unwrapped);
+                    } catch (ConfigException.Missing e) {
+                        // If the field is not marked as @{link Optional}...
+                        if (!isOptionalProperty(clazz, beanProp)) {
+                            // bubble the {@link Missing} exception up
+                            throw e;
+                        }
```

Comment:
Should this (sometimes?) be explicitly setting the field to `null`? I feel like if the config does not contain a setting, maybe we keep whatever the bean's constructor set, but if the config explicitly has `null` in it, we should probably explicitly set to `null`?